### PR TITLE
Let RetryPolicy throw exception when number of attempts exhausted

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
Remove the return value from attempt() so it either succeeds or throws exception
Add AttemptsExceededException to indicates number of attempts exhausted

The reason for this change is to prevent using RetryPolicy without checking the return value which might cause silent failure.

### CODE REVIEW

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/helix/HelixHelper.java [hunk 1]
```diff
@@ -70,67 +70,67 @@
...
-          return true;
         }
-      }
-    });
-
-    if (!successful) {
-      throw new RuntimeException("Failed to update ideal state for resource " + resourceName);
+      });
+    } catch (Exception e) {
+      throw new RuntimeException("Caught exception while updating ideal state for resource: " + resourceName, e);
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/helix/HelixHelper.java]: Correct me if I am wrong, but the earlier piece of code logs the actual exception encountered during attempts, whereas this one will only throw AttemptsExceededException ?

Comment (Jackie-Jiang) [reply pinot-common/src/main/java/com/linkedin/pinot/common/utils/helix/HelixHelper.java]: This one will log the actual exception encountered during attempts, or AttemptsExceededException only if no exception encountered but run out of attempts

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/retry/RetryPolicy.java [hunk 2]
```diff
@@ -22,12 +22,13 @@
...
+   * Attempts to do the operation until it succeeds, aborting if an exception is thrown by the operation or number of
+   * attempts exhausted.
    *
    * @param operation The operation to attempt, which returns true on success and false on failure.
-   * @return true if the operation succeeded or false if the operation did not succeed within the retries specified by
-   * this retry policy
+   * @throws Exception
    */
-  boolean attempt(Callable<Boolean> operation);
+  void attempt(Callable<Boolean> operation) throws Exception;
```
Comment (jfim) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/retry/RetryPolicy.java]: This should throw a more specific type of exception than `Exception` (eg. `AttemptsExceededException`, etc.).

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/retry/AttemptsExceededException.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.common.utils.retry;
+
+/**
+ * The <code>AttemptsExceededException</code> indicates that the operation does not succeed within maximum number of
+ * attempts.
+ */
+public class AttemptsExceededException extends Exception {
```
Comment (jfim) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/retry/AttemptsExceededException.java]: Should probably have a parent exception class like `AttempedOperationFailureException` or something similar.

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/retry/RetryPolicy.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/com/linkedin/pinot/common/utils/retry/RetryPolicy.java]: This could also throw arbitrary exceptions from the Callable

**Pull request merged**